### PR TITLE
Set up server, database, and auth

### PR DESCRIPTION
**Pull request by lwu15**
Finished setting up the server, database, and authentication. Currently, the database URI and authentication tokens are hardcoded. They will need to be updated accordingly when in production mode. The isLoggedIn flag is set to true for development, so that you won't be logged out when you make changes to the code.

### CODE REVIEW

File: main/models/db.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
+const { Pool } = require('pg');
+// require('dotenv').config();
+
+// For saving the database URI in the env file instead:
+// const pool = new Pool({ connectionString: process.env.DB_URI });
+
+const pool = new Pool({
+  connectionString:
+    'postgres://gmasbmxl:R44JPMUT3LjuZqZ2Hcmn2A65P4k2Lv-B@berry.db.elephantsql.com/gmasbmxl',
```
Comment (echochs1) [inline main/models/db.js]: can we put this in the .env?

File: main/controllers/authControllers.js [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        status: 400,
+        message: { error: 'Account with this email already exists.' },
+      });
+    }
+
+    let signupQuery, signupValues;
+
+    const hashedPassword = await bcrypt.hash(user_password, 10);
+
+    // Different queries to insert new user info depending on whether user has provided last name
```
Comment (echochs1) [inline main/controllers/authControllers.js:40]: why is this necessary?

File: main/controllers/authControllers.js [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+};
+
+// Verify user when login
+authControllers.verifyUser = async (req, res, next) => {
+  try {
+    const user_email = req.body.user_email;
+    const user_password = req.body.user_password;
+
+    // Find the user's info in the db
+    if (user_email && user_password) {
```
Comment (echochs1) [inline main/controllers/authControllers.js:73]: we can handle this conditional client side for efficiency

File: main/expressServer.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+app.use(express.urlencoded({ extended: true }));
+app.use(express.json());
+
+// TO DO: Update for production env / Generate a secret for the session
+const secret = process.env.SESSION_SECRET || crypto.randomBytes(64).toString('hex');
+
+// Inject the pool into the request object for easy access in routes
+app.use((req, res, next) => {
+  // console.log('Cookies:', req.cookies);
+  req.pool = pool;
```
Comment (echochs1) [inline main/expressServer.js]: do we want to expose the database pool in our entire server?

**Review by echochs1 — CHANGES_REQUESTED** [review]
Overall, looks pretty good. There are a few things that have to be changed before merging. I left some comments on those lines and would like to do a pair code review session.

File: src/app.jsx [hunk 5]
```diff
@@ -65,19 +67,46 @@ const theme = createTheme({
   },
 });
 
-function App() {
+const App = () => {
+  // TO DO: uncomment and set isLoggedIn to false for production setting
+  // const [isLoggedIn, setLoggedIn] = useState(false);
+
+  // For development mode, isLoggedIn is set to true
+  const [isLoggedIn, setLoggedIn] = useState(true);
```
Comment (echochs1) [inline src/app.jsx]: we can store sessions in the database to persist login information

File: src/app.jsx [hunk 6]
```diff
@@ -65,19 +67,46 @@ const theme = createTheme({
...
+          </>
+        )}
         <Routes>
-          <Route path="*" element={<Overview />} />
-          <Route path="/Connections" element={<Connections />} />
-          <Route path="/Alerts" element={<Alerts />} />
-          <Route path="/Brokers" element={<Brokers />} />
-          <Route path="/Producers" element={<Producers />} />
-          <Route path="/Consumers" element={<Consumers />} />
+          <Route path="/Login" element={<Login onLogin={handleLogin} />} />
```
Comment (echochs1) [inline src/app.jsx]: if logged in, hide login and sign up routes

File: main/expressServer.js (hunk 4 above)
Comment (echochs1) [reply main/expressServer.js]: no

**Comment by lwu15** [comment]
@echochs1
Elliot, thanks for the feedback and code review. I’ve pushed the following updates:
1. If already logged in, hid login and signup.
2. For the signup and login page, centered the component.
3. For login, can confirm that we ensure that both email and password are checked on the frontend and backend.
4. For signup, ensured that the password was at least 8 characters long, including a lowercase letter, an uppercase letter, and a number, before making the API call.
Note: Keeping the code for pooling in expressServer.js. The connection pool is actually defined in the ./models/db.js file. When using the db.query method to execute SQL queries in controllers, it internally calls pool.query, utilizing the connection pool for database interactions.
6. Created separate .env file and removed database uri from db.js. Sharing .env file with team.

In progress:
- Consider storing session information in the database (?).
- Work on removing signup and signin page margins.

File: main/models/db.js (hunk 1 above)
Comment (lwu15) [reply main/models/db.js]: updated

File: src/app.jsx (hunk 5 above)
Comment (lwu15) [reply src/app.jsx]: will look into it

File: main/models/db.js (hunk 1 above)
Comment (lwu15) [reply main/models/db.js]: @echochs1 as discussed, removed injection of pool to request object in expressServer.js

**Review by echochs1 — APPROVED** [review]
Looks great!
